dev_guide/cicd.adoc

```diff
@@ -0,0 +1,254 @@
+over reliable networks, and leverage *co-located* hardware ... somewhat analgous to a a single mirror, where something external synchronizes it with other mirrors, which are located in different geographies
+(cities, states, regions, countries, continents)
+GGM editorial note .... can a Cluster be even say building wide ???? can it span multiple floors, or are its performance needs too demanding??? ... are we too close to the capacity planning slippery slope
+
+Some implications of how you isolate:
+- As noted above with `*ImageStreams*`, promotion flows are possible both within a project and between projects serviced by the same OpenShift Internal Registry where tagging of Images trigger
+`*Deployments*` in separate `*Projects*` and/or `*Pods*`, where different teams in your organization perform different levels of validation against those different `*Pods*`.
+- But if computer processing or capactity concern reach the point that you can't share an OpenShift Master and Internal Registry across all your projects ...
+- Or  if you want to make sure *team A* can't accidently misconfigure resources for *team B*
+- Or if you want to make sure that say a problem with *application A* does not inadvertently impact running instances of *application B* ...
+- Or if your organization maintains separate versions of the requisite database(s) that service your application(s), and those versions are strictly isolated ....
+- Or if your application's network activity is as such that you want to isolate that activity at various stages of validation within your pipeline (say performance testing vs. integration testing) ...
+- then you isolate first along `*Project*` lines (where perhaps there is some duplication of API Objects) and the `*Cluster*` lines (where there it total duplication, and more like change in various connection settings to
+account for environment changes)
+
+More on this later when we dive into the more precise examples.
+
+
+[[upgrade-update]]
+== Upgrade and Update
+
+OpenShift `*Deployments*` are the mechanism for handling upgrade/update.
+
+`*Deployments*` have strategies that determine the process of deployment (how the older version is brought down, how the new version is brought up), and liveness and readiness probes (http requests or operating
+system level commands) for determining if the deployment is successful.
+
+Irrespective of how the deployment is triggered, the state of the deployment can be monitored from the OpenShift WebConsole, the OpenShift CLI, and in the case of `*CI/CD*` pipeline flows from Jenkins, from the
+Deployment related steps provide by the OpenShift Pipeline Plugin.  The OpenShift Pipeline Plugin has steps to verify if a `*Service*` is responsive, or if a given number or `*Replicas*` are up, or if a `*Deployment*`
```

Comment:
s/or if a given number/if a given number/
s/number or/number of/
s/Deployment in fact../Deployment has been triggered by an Image change./